code/modules/clothing/belts/polymorph_belt.dm

```diff
@@ -0,0 +1,152 @@
+	visible_message(span_notice("[user] begins scanning [target_mob] with [src]."))
+	if (!do_after(user, delay = 5 SECONDS, target = target_mob))
+		return
+	visible_message(span_notice("[user] scans [target_mob] with [src]."))
+	stored_mob_type = target_mob.type
+	update_transform_action()
+	playsound(src, 'sound/machines/ping.ogg', 50, FALSE)
+
+/// Make sure we can transform into the scanned target
+/obj/item/polymorph_belt/proc/update_transform_action()
+	if (isnull(stored_mob_type) || !active)
+		return
+	if (isnull(transform_action))
+		transform_action = add_item_action(/datum/action/cooldown/spell/shapeshift/polymorph_belt)
+	transform_action.update_type(stored_mob_type)
+
+/// Pre-activated polymorph belt
+/obj/item/polymorph_belt/functioning
+	active = TRUE
+	icon_state = "polybelt"
+	worn_icon_state = "polybelt"
+
+/// Ability provided by the polymorph belt
+/datum/action/cooldown/spell/shapeshift/polymorph_belt
+	name = "Invert Polymorphic Field"
+	cooldown_time = 30 SECONDS
+	school = SCHOOL_UNSET
+	invocation_type = INVOCATION_NONE
+	spell_requirements = NONE
+	possible_shapes = list(/mob/living/basic/cockroach)
+	/// Amount of time it takes us to transform back or forth
+	var/channel_time = 3 SECONDS
+
+/datum/action/cooldown/spell/shapeshift/polymorph_belt/before_cast(mob/living/cast_on)
+	. = ..()
+	if (. & SPELL_CANCEL_CAST)
+		return
+	if (channel_time <= 0)
+		return
+	if (DOING_INTERACTION_WITH_TARGET(cast_on, cast_on))
+		return . | SPELL_CANCEL_CAST
+
+	var/old_transform = cast_on.transform
+
+	var/animate_step = channel_time / 6
+	playsound(cast_on, 'sound/effects/wounds/crack1.ogg', 50)
+	animate(cast_on, transform = matrix() * 1.1, time = animate_step, easing = SINE_EASING)
+	animate(transform = matrix() * 0.9, time = animate_step, easing = SINE_EASING)
+	animate(transform = matrix() * 1.2, time = animate_step, easing = SINE_EASING)
+	animate(transform = matrix() * 0.8, time = animate_step, easing = SINE_EASING)
+	animate(transform = matrix() * 1.3, time = animate_step, easing = SINE_EASING)
+	animate(transform = matrix() * 0.1, time = animate_step, easing = SINE_EASING)
+
+	cast_on.balloon_alert(cast_on, "transforming...")
+	if (!do_after(cast_on, delay = channel_time, target = cast_on))
+		animate(cast_on, transform = matrix(), time = 0.25, easing = SINE_EASING)
+		cast_on.transform = old_transform
+		return . | SPELL_CANCEL_CAST
+	cast_on.visible_message(span_warning("[cast_on]'s body rearranges itself with a horrible crunching sound!"))
+	playsound(cast_on, 'sound/magic/demon_consume.ogg', 50, TRUE)
+
+/datum/action/cooldown/spell/shapeshift/polymorph_belt/after_cast(atom/cast_on)
+	. = ..()
+	if (QDELETED(owner))
+		return
+	animate(owner, transform = matrix() * 0.1, time = 0, easing = JUMP_EASING)
+	animate(transform = matrix(), time = 0.25 SECONDS, easing = SINE_EASING)
+
+/datum/action/cooldown/spell/shapeshift/polymorph_belt/Remove(mob/living/remove_from)
+	unshift_owner()
+	return ..()
```

Comment:
I'm surprised this works considering it `Remove`s then `Grant`s on cast (to the shapechange mob)

(Though this should probably be default behavior of shapechange)

Reply:
Oh you're right, I guess it works because while it is doing that, the buff has not finished being applied yet.